src/resolve-info.ts

```diff
@@ -0,0 +1,247 @@
+import {
+  doTypesOverlap,
+  FieldNode,
+  GraphQLCompositeType,
+  GraphQLError,
+  GraphQLInterfaceType,
+  GraphQLNamedType,
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLResolveInfo,
+  GraphQLSchema,
+  isCompositeType,
+  isListType,
+  isNonNullType,
+  isObjectType,
+  isUnionType,
+  SelectionSetNode
+} from "graphql";
+import memoize from "lodash.memoize";
+import deepMerge from "lodash.merge";
+import { ObjectPath } from "./ast";
+import { memoize2, memoize4 } from "./memoize";
+
+export interface GraphQLJitResolveInfo extends GraphQLResolveInfo {
+  fieldExpansion: FieldExpansion | true;
+}
+
+export interface FieldExpansion {
+  // The possible return types that the field can return
+  // It includes all the types in the Schema that intersect with the actual return type
+  [returnType: string]: TypeExpansion;
+}
+
+export interface TypeExpansion {
+  // The fields that are requested in the Query for a particular type
+  // `true` indicates a leaf node
+  [fieldName: string]: FieldExpansion | true;
+}
+
+/**
+ * Compute the GraphQLJitResolveInfo's `fieldExpansion` and return a function
+ * that returns the computed resolveInfo. This thunk is registered in
+ * context.dependencies for the field's resolveInfoName
+ */
+export function createResolveInfoThunk({
+  schema,
+  fragments,
+  operation,
+  parentType,
+  fieldName,
+  fieldType,
+  fieldNodes
+}: {
+  schema: GraphQLResolveInfo["schema"];
+  fragments: GraphQLResolveInfo["fragments"];
+  operation: GraphQLResolveInfo["operation"];
+  parentType: GraphQLObjectType;
+  fieldType: GraphQLOutputType;
+  fieldName: string;
+  fieldNodes: FieldNode[];
+}) {
+  const returnType = memoizedResolveEndType(fieldType);
+
+  // Result
+  const fieldExpansion: FieldExpansion = {};
+
+  if (returnType != null) {
+    for (const fieldNode of fieldNodes) {
+      deepMerge(
+        fieldExpansion,
+        memoizedExpandFieldNode(schema)(fragments)(fieldNode)(returnType)
```

Comment:
These 4 arguments memoizations do seem to have much value.

I expect `fieldNode` to be different.

Reply:
According to @boopathi , they look the same.